src/main/java/org/threadly/litesockets/UDPServer.java

```diff
@@ -90,8 +92,25 @@ public void close() {
     }
   }
   
-  protected int write(ByteBuffer bb, InetSocketAddress remoteAddress) throws IOException {
-    return channel.send(bb, remoteAddress);
+  protected int doWrite() {
+    WriteDataPair wdp = writeQueue.poll();
+    if(wdp != null) {
+      try {
+        return channel.send(wdp.bb, wdp.isa);
+      } catch (IOException e) {
+        return 0;
+      }
+    }
+    return 0;
+  }
+  
+  protected boolean needsWrite() {
+    return !writeQueue.isEmpty();
+  }
+  
+  protected void write(ByteBuffer bb, InetSocketAddress remoteAddress) throws IOException {
```

Comment:
Technically the TCPClient does not either, but it does give you futures to listen with to know when data is consumed.  I might add the same listenable to this and pass them through to the UDPclient